### ISSUE TITLE
Named format arguments introduce implicit positional arguments

### ISSUE THREAD

**Issue by Veetaha** [rust-lang/rust#93378]
I tried this code:

```rust
fn main() {
    let a = "a";
    let b = "b";
    
    println!("{a} {b} {} {} {c} {}", c = "c");
}
```

I expected to see a compile error that says that there are missing positional arguments or that it is prohibited to mix positional and named parameters.

Instead, this compiles and prints the following:
```
a b c a c b
```

This code is [available on the playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=91fcd6600665ef45dae25f2a735768fb).

### Meta

`rustc --version --verbose`:
```
rustc 1.58.1 (db9d1b20b 2022-01-20)
binary: rustc
commit-hash: db9d1b20bba1968c1ec1fc49616d4742c1725b4b
commit-date: 2022-01-20
host: x86_64-unknown-linux-gnu
release: 1.58.1
LLVM version: 13.0.0
```

**Comment by m-ou-se** [comment]
Oh dear. We should fix this quickly, before any code starts depending on this.

**Comment by xfix** [comment]
`println!("{c} {}", c = "c")` appears to be valid since Rust 1.12.

**Comment by m-ou-se** [comment]
> `println!("{c} {}", c = "c")` appears to be valid since Rust 1.12.

Yes, but that's not as bad, and arguably the right thing. The problem we should fix is the implicit captures getting used for `{}`.

**Comment by m-ou-se** [comment]
This should fix it: https://github.com/rust-lang/rust/pull/93394

**Comment by CAD97** [comment]
I just ran into this as well and reported #93415; I've tightened that one to just the fact that named arguments count as positional, so discussion on that can happen there.

**Comment by Nemo157** [comment]
Also see previously https://github.com/rust-lang/rust/issues/69110 and https://github.com/rust-lang/rust/issues/45256 for discussion about non-capture named arguments introducing positional arguments.

**Comment by camelid** [comment]
Assigning priority as [discussed](https://rust-lang.zulipchat.com/#narrow/stream/245100-t-compiler.2Fwg-prioritization.2Falerts/topic/.2393378.20Named.20format.20arguments.20introduce.20implicit.20positional.E2.80.A6) in the prioritization working group.

**Comment by joshtriplett** [comment]
We discussed this in today's @rust-lang/lang meeting, and we agreed that 1) we'd like this to stop being accepted, ASAP, and 2) we don't consider the PR fixing this to need T-lang review or approval.
